cmd/tsbs_generate_queries/databases/mongo/devops-naive.go

```diff
@@ -102,61 +93,263 @@ func (d *NaiveDevops) GroupByTimeAndPrimaryTag(qi query.Query, numMetrics int) {
+				},
+			},
+		},
+		{
+			"$group": bson.M{
+				"_id": bson.M{
+					"$dateTrunc": bson.M{"date": "$time", "unit": "hour"},
+				},
 			},
 		},
+		{
+			"$sort": bson.M{"_id": 1},
+		},
 	}
-	resultMap := group["$group"].(bson.M)
+	resultMap := pipelineQuery[1]["$group"].(bson.M)
 	for _, metric := range metrics {
-		resultMap["avg_"+metric] = bson.M{"$avg": "$fields." + metric}
+		resultMap["max_"+metric] = bson.M{"$max": "$" + metric}
 	}
-	pipelineQuery = append(pipelineQuery, group)
 
-	// Add sort operator
-	pipelineQuery = append(pipelineQuery, []bson.M{
-		{"$sort": bson.M{"_id.hostname": 1}},
-		{"$sort": bson.M{"_id.time": 1}},
-	}...)
-	pipelineQuery = append(pipelineQuery, bson.M{"$sort": bson.M{"_id.time": 1, "_id.hostname": 1}})
+	humanLabel := devops.GetMaxAllLabel("Mongo", nHosts)
+	q := qi.(*query.Mongo)
+	q.HumanLabel = []byte(humanLabel)
+	q.BsonDoc = pipelineQuery
+	q.CollectionName = []byte("point_data")
+	q.HumanDescription = []byte(fmt.Sprintf("%s: %s", humanLabel, interval.StartString()))
+}
+
+// HighCPUForHosts populates a query that gets CPU metrics when the CPU has high
+// usage between a time period for a number of hosts (if 0, it will search all hosts),
+// e.g. in pseudo-SQL:
+//
+// SELECT * FROM cpu
+// WHERE usage_user > 90.0
+// AND time >= '$TIME_START' AND time < '$TIME_END'
+// AND (hostname = '$HOST' OR hostname = '$HOST2'...)
+func (d *NaiveDevops) HighCPUForHosts(qi query.Query, nHosts int) {
+	interval := d.Interval.MustRandWindow(devops.HighCPUDuration)
+	metrics := devops.GetAllCPUMetrics()
 
-	humanLabel := devops.GetDoubleGroupByLabel("Mongo [NAIVE]", numMetrics)
+	pipelineQuery := []bson.M{}
+
+	// Must match in the documents that correspond to time, as well as optionally
+	// filter on those with the correct host if nHosts > 0
+	match := bson.M{
+		"$match": bson.M{
+			"measurement": "cpu",
+			"time": bson.M{
+				"$gte": interval.Start(),
+				"$lt":  interval.End(),
+			},
+		},
+	}
+	if nHosts > 0 {
+		hostnames, err := d.GetRandomHosts(nHosts)
+		panicIfErr(err)
+		matchMap := match["$match"].(bson.M)
+		matchMap["tags.hostname"] = bson.M{"$in": hostnames}
+	}
+	pipelineQuery = append(pipelineQuery, match)
+
+	project := bson.M{
+		"$project": bson.M{
+			"_id":  0,
+			"time": 1,
+			"tags": "$tags.hostname",
+		},
+	}
+	projectMap := project["$project"].(bson.M)
+	for _, metric := range metrics {
+		projectMap[metric] = 1
+	}
+	pipelineQuery = append(pipelineQuery, project)
+	pipelineQuery = append(pipelineQuery, bson.M{"$match": bson.M{"usage_user": bson.M{"$gt": 90.0}}})
+
+	humanLabel, err := devops.GetHighCPULabel("Mongo", nHosts)
+	panicIfErr(err)
 	q := qi.(*query.Mongo)
 	q.HumanLabel = []byte(humanLabel)
 	q.BsonDoc = pipelineQuery
 	q.CollectionName = []byte("point_data")
 	q.HumanDescription = []byte(fmt.Sprintf("%s: %s (%s)", humanLabel, interval.StartString(), q.CollectionName))
 }
+
+// LastPointPerHost finds the last row for every host in the dataset
```

Comment:
I _think_ the following might be a better/faster way to compute this:
```
{$match: {measurement: "cpu"}},
{$sort: {time: -1}},
{$group: {
   _id: "$tags.hostname",
   last_time: {$first: "$time"},
   // for each metric:
   metric: {$first: "$metric"},
}}
```

I am 50% sure I'm misunderstanding something though :D

Reply:
I for one am lost without the SQL to guide me!

Reply:
Thanks Katherine! We should continue brainstorming on that ticket if there's a better way - I'm believing now that it's a more complicated optimization than I think I was first imagining and laid out in the description. If we don't consider SERVER-55576 for 5.0 or anytime soon then maybe we should go back to the $lookup so our numbers are slightly less embarrassing... 40 seconds is rough :D

Reply:
Roger that! I can change this back to the previous query if it looks like SERVER-55576 won't be making it into 5.0.